tests/binary_agreement_mitm.rs

```diff
@@ -0,0 +1,506 @@
+    },
+    Stage {
+        source_groups: &[F],
+        dest_groups: &[A0, A1],
+        msg_type: MessageType::BVal,
+        msg_contents: Some(BoolFromState::AEstimated(true)),
+        msg_count: NODES_PER_GROUP * 2,
+    },
+    // !! Not mentioned in the GitHub issue, but seems necessary.
+    // F sends Aux(_) to A, because nodes in A need 2f+1 Aux messages
+    // before they broadcast their coins.
+    Stage {
+        source_groups: &[F],
+        dest_groups: &[A0, A1],
+        msg_type: MessageType::Aux,
+        msg_contents: Some(BoolFromState::AEstimated(false)),
+        msg_count: NODES_PER_GROUP * 2,
+    },
+    // Now all nodes in A broadcast their threshold shares over the coin,
+    // so since |A|+|F|=2f+1, the adversary can construct the random coin value s.
+    Stage {
+        source_groups: &[A0, A1],
+        dest_groups: &[F],
+        msg_type: MessageType::Coin,
+        msg_contents: None,
+        msg_count: NODES_PER_GROUP * 2,
+    },
+    // The nodes in F send BVAL(\neg s) to all the nodes in B,
+    // and all the BVAL(\neg s) messages from nodes in A are delivered to all nodes in B.
+    // Thus all the nodes in B broadcast AUX(\neg s).
+    Stage {
+        source_groups: &[A0, A1, F],
+        dest_groups: &[B],
+        msg_type: MessageType::BVal,
+        msg_contents: Some(BoolFromState::CoinValue(true)),
+        msg_count: (NODES_PER_GROUP * 2 + 1) * NODES_PER_GROUP,
+    },
+    // Deliver all AUX(\neg s) messages; there are 2f+1 of them,
+    // since either every node in A0 broadcast AUX(\neg s)
+    // or every node in A1 broadcast AUX(\neg s).
+    // Thus all nodes in B see 2f+1 AUX(\neg s) messages
+    // and get to the end of the round with bin_values=\neg s.
+    // Thus the nodes in B continue to the next round voting \neg s
+    // while the nodes in A continue to the next round voting s.
+    Stage {
+        source_groups: &[A0, A1, B, F],
+        dest_groups: &[B],
+        msg_type: MessageType::Aux,
+        msg_contents: Some(BoolFromState::CoinValue(true)),
+        msg_count: (NODES_PER_GROUP + 1) * (NODES_PER_GROUP)
+            + (NODES_PER_GROUP * (NODES_PER_GROUP - 1)),
+    },
+    // At this point all messages from the round are delivered, and the process repeats.
+];
+
+/// An adversary for the reordering attack.
+/// Described here: https://github.com/amiller/HoneyBadgerBFT/issues/59#issue-310368284
+/// Excluding the first node, which is F,
+/// A0 is the first third of nodes, A1 is the second third, and the rest are B.
+struct AbaCommonCoinAdversary {
+    stage: usize,
+    stage_progress: usize,
+    sent_stage_messages: bool,
+    epoch: u32,
+    coin_state: CoinState<NodeId>,
+    /// The estimated value for nodes in A.
+    a_estimated: bool,
+    // TODO this is really hacky but there's no better way to get this value
+    netinfo_mutex: Arc<Mutex<Option<Arc<NetworkInfo<NodeId>>>>>,
+}
+
+const NODES_PER_GROUP: usize = 2;
```

Comment:
Please randomize the number of nodes in a group. The attack targets groups of any size. The tests might try a few sizes in a range `[1..max_size]` for some `max_size`.

Reply:
On a second thought, no randomization would replace formal analysis because it would still generate finitely many tests. Since this is a regression test for a particular message ordering, a constant group size suffices here.

Reply:
FYI, in my testing, I never noticed group sizes having an effect on how the attack worked.